### PR TITLE
Throw when connection is closed and we try to execute commands

### PR DESCRIPTION
**Pull request by mariano54**
### Description

I ran into an issue where I had an asyncio method calling fetchall() on a Cursor, but it hung forever. This is because the connection was closed, and the loop stopped running so the fetchall future was never pulled out of the queue. An example test case is added.

### Fix

This PR changes _execute so it checks for whether the connection is open or not. _connect also has to be changed, because it used to call _execute (which would throw if there is no connection yet). I'm assuming all the other calls to _execute rely on the connection being open. Is this the right approach? If there is a better way to do it, I can implement it. 

Also, after the connection is closed, all futures in the queue are processed anyway.

### CODE REVIEW

**Comment by hoffmang9** [comment]
Is there anything we can further do to get this merged?

File: aiosqlite/tests/smoke.py [hunk 1]
```diff
@@ -354,6 +354,31 @@ async def test_iterdump(self):
...
 
+    async def test_cursor_on_closed_connection(self):
+        db = await aiosqlite.connect(TEST_DB)
+
+        cursor = await db.execute("select 1, 2")
+        await db.close()
+        with self.assertRaises(Exception):
+            await cursor.fetchall()
+        with self.assertRaises(Exception):
+            await cursor.fetchall()
```
Comment (jreese) [inline aiosqlite/tests/smoke.py:365]: Would prefer to see something like `with self.assertRaisesRegex(ValueError, "connection closed")` to also ensure this catches the specific exception it's expecting to catch.

File: aiosqlite/tests/smoke.py [hunk 2]
```diff
@@ -354,6 +354,31 @@ async def test_iterdump(self):
...
+        tasks = []
+        for i in range(100):
+            if i == 50:
+                tasks.append(asyncio.ensure_future(db.close()))
+            tasks.append(asyncio.ensure_future(cursor.fetchall()))
+        for task in tasks:
+            try:
+                await task
+            except Exception:
+                pass
```
Comment (jreese) [inline aiosqlite/tests/smoke.py:380]: Is there a way to make sure the tasks are returning (or throwing) the expected values/exceptions?

**Review by jreese — CHANGES_REQUESTED** [review]
Functionality looks good, but I'd like to see better test cases verifying that functionality.

**Comment by mariano54** [comment]
Would you mind taking another look and merging? it would help us out to have this included in a release.

Thanks :)

**Comment by hoffmang9** [comment]
Checking in on this. Is there anything further you need from @mariano54 ?

**Comment by jreese** [comment]
Sorry for the delay on returning to this. Thank you for the fixes!

File: aiosqlite/tests/smoke.py [hunk 3]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
                 ],
             )
 
+    async def test_cursor_on_closed_connection(self):
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:394]: async def test_cursor_on_closed_connection(self):

File: aiosqlite/tests/smoke.py [hunk 4]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        db = await aiosqlite.connect(TEST_DB)
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:395]: db = await aiosqlite.connect(TEST_DB)

File: aiosqlite/tests/smoke.py [hunk 5]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+
+        cursor = await db.execute("select 1, 2")
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:397]: cursor = await db.execute("select 1, 2")

File: aiosqlite/tests/smoke.py [hunk 6]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        await db.close()
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:398]: await db.close()

File: aiosqlite/tests/smoke.py [hunk 7]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        with self.assertRaisesRegex(ValueError, "Connection closed"):
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:399]: with self.assertRaisesRegex(ValueError, "Connection closed"):

File: aiosqlite/tests/smoke.py [hunk 8]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+            await cursor.fetchall()
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:400]: await cursor.fetchall()

File: aiosqlite/tests/smoke.py [hunk 9]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        with self.assertRaisesRegex(ValueError, "Connection closed"):
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:401]: with self.assertRaisesRegex(ValueError, "Connection closed"):

File: aiosqlite/tests/smoke.py [hunk 10]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+            await cursor.fetchall()
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:402]: await cursor.fetchall()

File: aiosqlite/tests/smoke.py [hunk 11]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+
+    async def test_cursor_on_closed_connection_loop(self):
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:404]: async def test_cursor_on_closed_connection_loop(self):

File: aiosqlite/tests/smoke.py [hunk 12]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        db = await aiosqlite.connect(TEST_DB)
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:405]: db = await aiosqlite.connect(TEST_DB)

File: aiosqlite/tests/smoke.py [hunk 13]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+
+        cursor = await db.execute("select 1, 2")
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:407]: cursor = await db.execute("select 1, 2")

File: aiosqlite/tests/smoke.py [hunk 14]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        tasks = []
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:408]: tasks = []

File: aiosqlite/tests/smoke.py [hunk 15]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        for i in range(100):
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:409]: for i in range(100):

File: aiosqlite/tests/smoke.py [hunk 16]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+            if i == 50:
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:410]: if i == 50:

File: aiosqlite/tests/smoke.py [hunk 17]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+                tasks.append(asyncio.ensure_future(db.close()))
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:411]: tasks.append(asyncio.ensure_future(db.close()))

File: aiosqlite/tests/smoke.py [hunk 18]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+            tasks.append(asyncio.ensure_future(cursor.fetchall()))
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:412]: tasks.append(asyncio.ensure_future(cursor.fetchall()))

File: aiosqlite/tests/smoke.py [hunk 19]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+        for task in tasks:
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:413]: for task in tasks:

File: aiosqlite/tests/smoke.py [hunk 20]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+            try:
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:414]: try:

File: aiosqlite/tests/smoke.py [hunk 21]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+                await task
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:415]: await task

File: aiosqlite/tests/smoke.py [hunk 22]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+            except sqlite3.ProgrammingError:
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:416]: except sqlite3.ProgrammingError:

File: aiosqlite/tests/smoke.py [hunk 23]
```diff
@@ -391,6 +391,31 @@ async def test_iterdump(self):
...
+                pass
```
Comment (KolomboPulse) [inline aiosqlite/tests/smoke.py:417]: pass

File: aiosqlite/core.py [hunk 24]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
 
         :meta private:
         """
-        while self._running:
```
Comment (KolomboPulse) [inline aiosqlite/core.py:90]: while self._running:

File: aiosqlite/core.py [hunk 25]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
...
+        while True:
```
Comment (KolomboPulse) [inline aiosqlite/core.py:90]: while True:

File: aiosqlite/core.py [hunk 26]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
...
+            # Continues running until all queue items are processed,
```
Comment (KolomboPulse) [inline aiosqlite/core.py:91]: # Continues running until all queue items are processed,

File: aiosqlite/core.py [hunk 27]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
...
+            # even after connection is closed (so we can finalize all
```
Comment (KolomboPulse) [inline aiosqlite/core.py:92]: # even after connection is closed (so we can finalize all

File: aiosqlite/core.py [hunk 28]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
...
+            # futures)
```
Comment (KolomboPulse) [inline aiosqlite/core.py:93]: # futures)

File: aiosqlite/core.py [hunk 29]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
...
             try:
                 future, function = self._tx.get(timeout=0.1)
             except Empty:
-                continue
-
+                if self._running:
```
Comment (KolomboPulse) [inline aiosqlite/core.py:97]: if self._running:

File: aiosqlite/core.py [hunk 30]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
...
+                    continue
```
Comment (KolomboPulse) [inline aiosqlite/core.py:98]: continue

File: aiosqlite/core.py [hunk 31]
```diff
@@ -87,12 +87,16 @@ def run(self) -> None:
...
+                break
```
Comment (KolomboPulse) [inline aiosqlite/core.py:99]: break

File: aiosqlite/core.py [hunk 32]
```diff
@@ -114,6 +118,9 @@ def set_exception(fut, e):
 
     async def _execute(self, fn, *args, **kwargs):
         """Queue a function with the given arguments for execution."""
+        if not self._running or not self._connection:
```
Comment (KolomboPulse) [inline aiosqlite/core.py:121]: if not self._running or not self._connection:

File: aiosqlite/core.py [hunk 33]
```diff
@@ -114,6 +118,9 @@ def set_exception(fut, e):
...
+            raise ValueError("Connection closed")
```
Comment (KolomboPulse) [inline aiosqlite/core.py:122]: raise ValueError("Connection closed")

File: aiosqlite/core.py [hunk 34]
```diff
@@ -125,7 +132,9 @@ async def _connect(self) -> "Connection":
         """Connect to the actual sqlite database."""
         if self._connection is None:
             try:
-                self._connection = await self._execute(self._connector)
```
Comment (KolomboPulse) [inline aiosqlite/core.py:128]: self._connection = await self._execute(self._connector)

File: aiosqlite/core.py [hunk 35]
```diff
@@ -125,7 +132,9 @@ async def _connect(self) -> "Connection":
...
+                future = asyncio.get_event_loop().create_future()
```
Comment (KolomboPulse) [inline aiosqlite/core.py:135]: future = asyncio.get_event_loop().create_future()

File: aiosqlite/core.py [hunk 36]
```diff
@@ -125,7 +132,9 @@ async def _connect(self) -> "Connection":
...
+                self._tx.put_nowait((future, self._connector))
```
Comment (KolomboPulse) [inline aiosqlite/core.py:136]: self._tx.put_nowait((future, self._connector))

File: aiosqlite/core.py [hunk 37]
```diff
@@ -125,7 +132,9 @@ async def _connect(self) -> "Connection":
...
+                self._connection = await future
```
Comment (KolomboPulse) [inline aiosqlite/core.py:137]: self._connection = await future

**Pull request merged**